sparkle/instance/instances.py

```diff
@@ -48,26 +27,125 @@ def size(self: InstanceSet) -> int:
+            self._instance_names = [p.name for p in self._instance_paths]
+
+    @property
+    def name(self: FileInstanceSet) -> str:
+        """Get instance set name."""
+        return self._name
+
+
+class MultiFileInstanceSet(InstanceSet):
+    """Object representation of a set of multi-file instances."""
+    instance_csv = "instances.csv"
+
+    def __init__(self: MultiFileInstanceSet, target: Path | list[str, Path]) -> None:
+        """Initialise an Instances object from a directory.
+
+        Args:
+            target: Path to the instances directory. Will read from instance_list.csv.
+                If directory is a list of [str, Path], create an Instance set of one.
+        """
+        super().__init__(target)
+        if isinstance(target, list):
+            # A single instance represented as a list of [name, path1, path2, ...]
+            instance_list = target
+            target = target[1].parent
+        elif isinstance(target, Path):
+            # A path pointing to the directory of instances
+            instance_file = target / MultiFileInstanceSet.instance_csv
+            # Read from instance_file
+            instance_list = [line for line in csv.reader(instance_file.open())]
+
+        self.directory = target if target.is_dir() else target.parent
+        self._instance_names, self._instance_paths = [], []
+        for instance in instance_list:
+            self._instance_names.append(instance[0])
+            self._instance_paths.append([
+                (self.directory / f) if isinstance(f, str) else f for f in instance[1:]])
+
+    @property
+    def all_paths(self: MultiFileInstanceSet) -> list[Path]:
+        """Returns all file paths in the instance set as a flat list."""
+        return [p for instance in self._instance_paths for p in instance] + [
+            self.directory / MultiFileInstanceSet.instance_csv]
+
+
+class IterableFileInstanceSet(InstanceSet):
+    """Object representation of files containing multiple instances."""
+    supported_filetypes = set([".csv", ".npy"])
+
+    def __init__(self: IterableFileInstanceSet, target: Path) -> None:
+        """Initialise an InstanceSet from a single file.
+
+        Args:
+            target: Path to the instances directory. If multiple files are found,
+                they are assumed to have the same number of instances.
+        """
+        super().__init__(target)
+        self.directory = target
+        self._instance_paths =\
+            [p for p in self.directory.iterdir()
+             if p.suffix in IterableFileInstanceSet.supported_filetypes]
+        self._size = len(self._instance_paths[0].open().readlines())
+        self._instance_names = [p.name for p in self._instance_paths]
+
+    @property
+    def size(self: IterableFileInstanceSet) -> int:
+        """Returns the number of instances in the set."""
+        return self._size
+
+    @staticmethod
+    def __determine_size__(file: Path) -> int:
```

Comment:
is this an actual existing method? Why not use `__len__`? then we can do `len(instanceset)`